Fix L4-only policy egress to world and CIDR-only egress to world

joestringer opened the pull request:
Simplify and streamline the policy lookup in the egress path, to make the lookup respect entity-based policies such as those matching "cluster" or "world", to allow rules with L3 CIDR matches with L4 wildcard to work, and to allow rules with L4 match and an L3 wildcard to work for traffic destined outside the cluster. This should also improve the scalability of L4 rules for CIDR traffic by removing the embedded L4 lookup.

Fixes: #3371 
Fixes: #3426 
Fixes: #3427

joestringer committed:
bpf: Apply CIDR egress policy earlier

Adjust the application of L4 / CIDR egress policy to allow traffic that
matches either the L4 policy or the CIDR policy, as part of the same
check. Previously, if L4-only policy allowed traffic, it would be denied
by CIDR; and if CIDR-only policy allowed traffic, it would be denied by
L4. These cases should now work as expected.

Fixes: #3371
Fixes: #3427

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Block ICMP at L4 when policy is enabled

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -516,7 +512,7 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb)
 		verdict = l4_policy_lookup(skb, tuple.nexthdr, tuple.dport,
 					   CT_EGRESS, false);
-		if (verdict < 0)
+		if (verdict < 0 && unlikely(!lpm4_egress_lookup(orig_dip)))
```
This seems wrong

```
if (verdict < 0)
    return verdict
else if unlikely(!lpm4_egress_lookup(orig_dip))
    return DROP_POLICY_CIDR;
```
?

joestringer replied:
Here's the policy I was thinking about from this perspective:

```
[                                                                               
  {                                                                             
    "endpointSelector": {                                                       
      "matchLabels": {                                                          
        "container:id": "app2"                                                  
      }                                                                         
    },                                                                          
    "egress": [                                                                 
      {                                                                         
        "toCIDRSet": [                                                          
          {                                                                     
            "cidr": "8.8.0.0/16"                                                
          }                                                                     
        ]                                                                       
      },                                                                                                                                                                                 
      {                                                                         
         "toPorts": [                                                            
          {"ports":[ {"port": "80", "protocol": "TCP"} ] }                      
        ]                                                                       
      }                                                                         
    ]                                                                           
  }                                                                             
] 
```

I expected that with this policy, I can make a DNS request to 8.8.8.8 (ie port 53 within the CIDR), then make a web request for a website (port 80 which matches the L4 rule). This is a similar but simplified version of the policy in #3371.

With the PR as of now, this works, though I haven't gone through all of the existing policy tests to see if I've missed something obvious that it breaks.

tgraf replied:
I don't understand the `verdict < 0 && !lpm4_egress_lookup(orig_dip)` part. I read this as, we only ever apply the CIDR policy if L4 disallows it. If no L4 policy exists, it will return 0 and we never apply the CIDR rules. No?

tgraf replied:
I think we need to extend the `func (r Rule) resolveCIDR()` code to to add a `0/0` on an L4 only rule to whitelist egress CIDR.

joestringer replied:
The `verdict < 0 && !lpm_egress_lookup(orig_dip)` code is trying to indicate `if (l4_policy_drops_packet() && cidr_policy_drops_packet()) { return DROP; }` - the implication being that in the inverse case, if either policy allows the packet, then it should be let through.

The "no L4 policy exists" case is why I introduced the `l4_egress_policy()` changes in that patch - it says that if there's no L4 policy, but there are other egress policies, then from an L4 perspective it should be dropped; however it could still be allowed by a CIDR allow rule.

I prepared the patch for the latter `resolveCIDR()` fix, however it doesn't solve #3371 because the packet will be dropped by L4 policy before it gets to the CIDR check. If my interpretation of the intent of the above policy is wrong, then I agree that's the right solution and I can submit it.

tgraf replied:
Thanks, this helps understand the intent. I think you are right. I was worried about a case where L4 would allow it when CIDR would not but I think you are right that it can't happen.

We do not allow support toCIDR && toPorts which implies that any CIDR rule *always* implies L4 should not even apply.

We still want to allow proxy port though for certain parts. We have the following option that looks simple and clean:
```
static inline int policy_can_egress4(struct __sk_buff *skb,
                                     struct ipv4_ct_tuple *tuple)
{
        struct remote_endpoint_info *info;
        __u16 identity = 0;
        __be32 daddr;

        daddr = ipv4_ct_tuple_get_daddr(tuple);
        info = lookup_ip4_remote_endpoint(daddr);
        if (info)
                identity = info->sec_label;
        cilium_dbg(skb, info ? DBG_IP_ID_MAP_SUCCEED4 : DBG_IP_ID_MAP_FAILED4,
                   daddr, identity);

        return policy_can_egress(skb, identity, tuple->dport, tuple->nexthdr);
}
```

We change the above to default to `world` (2) instead of `0` which means that a user could define the following and it will work correctly:

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "allow-from-localhost"
spec:
  endpointSelector:
    matchLabels:
      id: service1
  egress:
    - toEntities:
      - world
      toPorts:
      - ports:
        - port: "80"
          protocol: TCP
```

It will also work with:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "allow-from-localhost"
spec:
  endpointSelector:
    matchLabels:
      id: service1
  egress:
    - toPorts:
      - ports:
        - port: "80"
          protocol: TCP
```

.. as the missing toEndpoints should result in a L3+L4 with world.

tgraf replied:
We would not even have to perform any L4 check for the non label case. Either L4 is defined in the identity based lookup or it's L4 allow all.

joestringer replied:
I was actually wondering about setting this default properly according to whether it's in the cluster or not, and if not, use world. I think it would fit well with this, and would allow applying policies based on things you know are within the cluster but aren't (currently) managed by cilium.

tgraf commented on bpf/lib/l4.h:
```diff
@@ -249,6 +249,10 @@ l4_policy_lookup(struct __sk_buff *skb, __u8 nh, __be16 dport, int dir,
 			cilium_dbg(skb, DBG_L4_POLICY, proxy_port, CT_INGRESS);
+#ifdef POLICY_INGRESS
+		} else {
+			proxy_port = DROP_POLICY_L4;
```
👍

joestringer committed:
bpf: Perform external policy lookup via POLICY_MAP

By default, use CLUSTER_ID / WORLD_ID for the egress policy lookup,
depending on the destination of the packet. If the destination identity
is not known, these will look up the policy for the entity which allows
L4 policy to be applied as well.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Reduce IP comparisons

These IP comparisons are performed earlier in the egress path and are
cached, reuse the cached value.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Remove embedded L4 policy map

L4 policies applying to traffic outside the cluster IP range can only be
applied to all such traffic via either no L3 match, or an L3 match on
the world entity. We already push such policies down into the POLICY_MAP
and look them up in policy_can_egress*(), so we can reuse this rather
than using the embedded L4 policy array. This allows us to get rid of
the embedded L4 policy for such cases, which simplifies the datapath and
expands the scalability of L4 policies.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf reviewed (approved):
Great work, thanks!
